packages/flutter/lib/src/material/magnifier.dart

```diff
@@ -0,0 +1,331 @@
+    shouldDisplayHandlesInMagnifier: defaultTargetPlatform == TargetPlatform.iOS,
+    magnifierBuilder: (
+      BuildContext context,
+      MagnifierController controller,
+      ValueNotifier<MagnifierOverlayInfoBearer> magnifierSelectionOverlayInfoBearer,
+    ) {
+      switch (defaultTargetPlatform) {
+        case TargetPlatform.iOS:
+          return CupertinoTextMagnifier(
+            controller: controller,
+            magnifierOverlayInfoBearer: magnifierSelectionOverlayInfoBearer,
+          );
+        case TargetPlatform.android:
+          return TextMagnifier(
+              magnifierSelectionOverlayInfoBearer: magnifierSelectionOverlayInfoBearer,
+          );
+        case TargetPlatform.fuchsia:
+        case TargetPlatform.linux:
+        case TargetPlatform.macOS:
+        case TargetPlatform.windows:
+          return null;
+      }
+    }
+  );
+
+  /// The duration that the position is animated if [TextMagnifier] just switched
+  /// between lines.
+  @visibleForTesting
+  static const Duration jumpBetweenLinesAnimationDuration =
+      Duration(milliseconds: 70);
+
+  /// [TextMagnifier] positions itself based on [magnifierSelectionOverlayInfoBearer].
+  ///
+  /// {@macro widgets.material.magnifier.positionRules}
+  final ValueNotifier<MagnifierOverlayInfoBearer>
+      magnifierSelectionOverlayInfoBearer;
+
+  @override
+  State<TextMagnifier> createState() => _TextMagnifierState();
+}
+
+class _TextMagnifierState extends State<TextMagnifier> {
+  // Should _only_ be null on construction. This is because of the animation logic.
+  //
+  // Animations are added when `last_build_y != current_build_y`. This condition
+  // is true on the inital render, which would mean that the inital
+  // build would be animated - this is undesired. Thus, this is null for the
+  // first frame and the condition becomes `magnifierPosition != null && last_build_y != this_build_y`.
+  Offset? _magnifierPosition;
+
+  // A timer that unsets itself after an animation duration.
+  // If the timer exists, then the magnifier animates its position -
+  // if this timer does not exist, the magnifier tracks the gesture (with respect
+  // to the positioning rules) directly.
+  Timer? _positionShouldBeAnimatedTimer;
+  bool get _positionShouldBeAnimated => _positionShouldBeAnimatedTimer != null;
+
+  Offset _extraFocalPointOffset = Offset.zero;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.magnifierSelectionOverlayInfoBearer
+        .addListener(_determineMagnifierPositionAndFocalPoint);
+  }
+
+  @override
+  void dispose() {
+    widget.magnifierSelectionOverlayInfoBearer
+        .removeListener(_determineMagnifierPositionAndFocalPoint);
+
+    if (_positionShouldBeAnimatedTimer != null) {
+      _positionShouldBeAnimatedTimer!.cancel();
+    }
+
+    super.dispose();
+  }
+
+  @override
+  void didChangeDependencies() {
+    _determineMagnifierPositionAndFocalPoint();
```

Comment:
since `_determineMagnifierPositionAndFocalPoint` calls `setState`, wouldn't it assert if it gets called here? Also what's the reason for calling this in `didChangeDependencies`?

Reply:
We use mediaQuery.of(context).size in the function call. This would throw if it was called from initState. It is safe to call initState from didChangeDependencies